From 1ab21e7edd748ac00f59b5aa5c2ae4e41beb4843 Mon Sep 17 00:00:00 2001
From: Jing Zhao <jing9@apache.org>
Date: Wed, 19 Aug 2015 10:36:56 -0700
Subject: [PATCH 0773/1023] HDFS-8867. Enable optimized block reports.
 Contributed by Daryn Sharp.

(cherry picked from commit f61120d964a609ae5eabeb5c4d6c9afe0a15cad8)
(cherry picked from commit c0a4cd978a9142a1fb9f6367d9af0dc7f0e711d9)
(cherry picked from commit 77feee66b20b88d3cd854c11bdc25ed9ca25a0c0)
---
 .../hadoop/hdfs/server/protocol/NamespaceInfo.java |   17 +++++++++++++----
 .../hadoop/hdfs/protocol/TestBlockListAsLongs.java |    9 ++++++++-
 2 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/protocol/NamespaceInfo.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/protocol/NamespaceInfo.java
index a7439a0..caca646 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/protocol/NamespaceInfo.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/protocol/NamespaceInfo.java
@@ -47,18 +47,27 @@
 
   // only authoritative on the server-side to determine advertisement to
   // clients.  enum will update the supported values
-  private static long CAPABILITIES_SUPPORTED = 0;
+  private static final long CAPABILITIES_SUPPORTED = getSupportedCapabilities();
+
+  private static long getSupportedCapabilities() {
+    long mask = 0;
+    for (Capability c : Capability.values()) {
+      if (c.supported) {
+        mask |= c.mask;
+      }
+    }
+    return mask;
+  }
 
   public enum Capability {
     UNKNOWN(false),
     STORAGE_BLOCK_REPORT_BUFFERS(true); // use optimized ByteString buffers
+    private final boolean supported;
     private final long mask;
     Capability(boolean isSupported) {
+      supported = isSupported;
       int bits = ordinal() - 1;
       mask = (bits < 0) ? 0 : (1L << bits);
-      if (isSupported) {
-        CAPABILITIES_SUPPORTED |= mask;
-      }
     }
     public long getMask() {
       return mask;
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/protocol/TestBlockListAsLongs.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/protocol/TestBlockListAsLongs.java
index 9ead765..bf29373 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/protocol/TestBlockListAsLongs.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/protocol/TestBlockListAsLongs.java
@@ -187,7 +187,14 @@ private void checkReplicas(Map<Long,Replica> expectedReplicas,
     }
     assertTrue(reportReplicas.isEmpty());
   }
-  
+
+  @Test
+  public void testCapabilitiesInited() {
+    NamespaceInfo nsInfo = new NamespaceInfo();
+    assertTrue(
+        nsInfo.isCapabilitySupported(Capability.STORAGE_BLOCK_REPORT_BUFFERS));
+  }
+
   @Test
   public void testDatanodeDetect() throws ServiceException, IOException {
     final AtomicReference<BlockReportRequestProto> request =
-- 
1.7.9.5

